File: VDFEditor/src/test/java/de/treichels/hott/vdfeditor/actions/UndoBufferTest.java
package de.treichels.hott.vdfeditor.actions;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.List;

import org.junit.Before;
import org.junit.Test;

public class UndoBufferTest {
    private List<String> items;
    private UndoBuffer<String> buffer;

    @Before
    public void setup() {
        items = new ArrayList<>();
        buffer = new UndoBuffer<>(items);
    }

    @Test
    public void testClear() {
        testPushAdd();
        buffer.clear();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(0, buffer.getIndex());
        assertEquals(0, buffer.getActions().size());
    }

    @Test
    public void testEmpty() {
        assertEquals("[]", items.toString());
        assertEquals(0, buffer.getIndex());
        assertEquals(0, buffer.getActions().size());
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testEmptyRedo() {
        buffer.redo();
    }

    @Test(expected = IndexOutOfBoundsException.class)
    public void testEmptyRedo2() {
        testPushAdd();
        buffer.redo();
    }

    @Test(expected = ArrayIndexOutOfBoundsException.class)
    public void testEmptyUndo() {
        buffer.undo();
    }

    @Test(expected = ArrayIndexOutOfBoundsException.class)
    public void testEmptyUndo2() {
        testUndoAdd();
        buffer.undo();
    }

    @Test
    public void testPop() {
        testPushAdd();

        buffer.pop();
        assertEquals("[a, b]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(2, buffer.getActions().size());

        buffer.pop();
        assertEquals("[a]", items.toString());
        assertEquals(1, buffer.getIndex());
        assertEquals(1, buffer.getActions().size());

        buffer.pop();
        assertEquals("[]", items.toString());
        assertEquals(0, buffer.getIndex());
        assertEquals(0, buffer.getActions().size());
    }

    @Test
    public void testPushAdd() {
        buffer.push(new AddAction<>("a"));
        assertEquals("[a]", items.toString());
        assertEquals(1, buffer.getIndex());
        assertEquals(1, buffer.getActions().size());

        buffer.push(new AddAction<>("b"));
        assertEquals("[a, b]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(2, buffer.getActions().size());

        buffer.push(new AddAction<>("c"));
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());
    }

    @Test
    public void testPushAfterUndo() {
        testPushAdd();

        buffer.undo();
        assertEquals("[a, b]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.push(new AddAction<>("d"));
        assertEquals("[a, b, d]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.undo();
        buffer.undo();
        assertEquals("[a]", items.toString());
        assertEquals(1, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.push(new AddAction<>("e"));
        assertEquals("[a, e]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(2, buffer.getActions().size());
    }

    @Test
    public void testPushInsert() {
        testPushAdd();

        buffer.push(new InsertAction<>(1, "x"));
        assertEquals("[a, x, b, c]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testPushMove() {
        testPushAdd();

        buffer.push(new MoveAction<>(1, 2));
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testPushMoveDown() {
        testPushAdd();

        buffer.push(new MoveDownAction<>(2));
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testPushMoveUp() {
        testPushAdd();

        buffer.push(new MoveUpAction<>(1));
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testPushRemove() {
        testPushAdd();

        buffer.push(new RemoveAction<>(1));
        assertEquals("[a, c]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testPushReplace() {
        testPushAdd();

        buffer.push(new ReplaceAction<>(1, "x"));
        assertEquals("[a, x, c]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testRedoAdd() {
        testUndoAdd();

        buffer.redo();
        assertEquals("[a]", items.toString());
        assertEquals(1, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.redo();
        assertEquals("[a, b]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.redo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());
    }

    @Test
    public void testRedoMove() {
        testUndoMove();
        buffer.redo();
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testRedoMoveDown() {
        testUndoMove();
        buffer.redo();
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testRedoMoveUp() {
        testUndoMove();
        buffer.redo();
        assertEquals("[a, c, b]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testRedoRemove() {
        testUndoRemove();
        buffer.redo();
        assertEquals("[a, c]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testRedoReplace() {
        testUndoReplace();
        buffer.redo();
        assertEquals("[a, x, c]", items.toString());
        assertEquals(4, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testUndoAdd() {
        testPushAdd();

        buffer.undo();
        assertEquals("[a, b]", items.toString());
        assertEquals(2, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.undo();
        assertEquals("[a]", items.toString());
        assertEquals(1, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());

        buffer.undo();
        assertEquals("[]", items.toString());
        assertEquals(0, buffer.getIndex());
        assertEquals(3, buffer.getActions().size());
    }

    @Test
    public void testUndoMove() {
        testPushMove();

        buffer.undo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testUndoMoveDown() {
        testPushMoveDown();

        buffer.undo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testUndoMoveUp() {
        testPushMoveUp();

        buffer.undo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testUndoRemove() {
        testPushRemove();

        buffer.undo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }

    @Test
    public void testUndoReplace() {
        testPushReplace();

        buffer.undo();
        assertEquals("[a, b, c]", items.toString());
        assertEquals(3, buffer.getIndex());
        assertEquals(4, buffer.getActions().size());
    }
}

File: VDFEditor/src/test/kotlin/de/treichels/hott/vdfeditor/actions/UndoBufferTest.kt
package de.treichels.hott.vdfeditor.actions

import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class UndoBufferTest {
    private val items = mutableListOf<String>()
    private var buffer = UndoBuffer(items)

    @Before
    fun setup() {
        items.clear()
        buffer.clear()
    }

    @Test
    fun testClear() {
        testPushAdd()

        buffer.clear()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(0, buffer.index.toLong())
        assertEquals(0, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertFalse(buffer.canUndo)
        assertFalse(buffer.canUndoProperty().get())
    }

    @Test
    fun testEmpty() {
        assertEquals("[]", items.toString())
        assertEquals(0, buffer.index.toLong())
        assertEquals(0, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertFalse(buffer.canUndo)
        assertFalse(buffer.canUndoProperty().get())
    }

    @Test
    fun testEmptyRedo() {
        testEmpty()

        buffer.redo()
        testEmpty()
    }

    @Test
    fun testEmptyRedo2() {
        testRedoAdd()

        buffer.redo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testEmptyUndo() {
        testEmpty()

        buffer.undo()
        testEmpty()
    }

    @Test
    fun testEmptyUndo2() {
        testUndoAdd()

        buffer.undo()
        assertEquals("[]", items.toString())
        assertEquals(0, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertFalse(buffer.canUndo)
        assertFalse(buffer.canUndoProperty().get())
    }

    @Test
    fun testPop() {
        testPushAdd()

        buffer.pop()
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(2, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.pop()
        assertEquals("[a]", items.toString())
        assertEquals(1, buffer.index.toLong())
        assertEquals(1, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.pop()
        testEmpty()
    }

    @Test
    fun testPushAdd() {
        testEmpty()

        buffer.push(AddAction("a"))
        assertEquals("[a]", items.toString())
        assertEquals(1, buffer.index.toLong())
        assertEquals(1, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.push(AddAction("b"))
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(2, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.push(AddAction("c"))
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushAfterUndo() {
        testPushAdd()

        buffer.undo()
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.push(AddAction("d"))
        assertEquals("[a, b, d]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a]", items.toString())
        assertEquals(1, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.push(AddAction("e"))
        assertEquals("[a, e]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(2, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushInsert() {
        testPushAdd()

        buffer.push(InsertAction(1, "x"))
        assertEquals("[a, x, b, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, x, b, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushMove() {
        testPushAdd()

        buffer.push(MoveAction(1, 2))
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushMoveDown() {
        testPushAdd()

        buffer.push(MoveDownAction(1))
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushMoveUp() {
        testPushAdd()

        buffer.push(MoveUpAction(2))
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushRemove() {
        testPushAdd()

        buffer.push(RemoveAction(1))
        assertEquals("[a, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testPushReplace() {
        testPushAdd()

        buffer.push(ReplaceAction(1, "x"))
        assertEquals("[a, x, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, x, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoAdd() {
        testUndoAdd()

        buffer.redo()
        assertEquals("[a]", items.toString())
        assertEquals(1, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.redo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoMove() {
        testUndoMove()

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoMoveDown() {
        testUndoMove()

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoMoveUp() {
        testUndoMove()

        buffer.redo()
        assertEquals("[a, c, b]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoRemove() {
        testUndoRemove()

        buffer.redo()
        assertEquals("[a, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testRedoReplace() {
        testUndoReplace()

        buffer.redo()
        assertEquals("[a, x, c]", items.toString())
        assertEquals(4, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertFalse(buffer.canRedo)
        assertFalse(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoAdd() {
        testPushAdd()

        buffer.undo()
        assertEquals("[a, b]", items.toString())
        assertEquals(2, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[a]", items.toString())
        assertEquals(1, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())

        buffer.undo()
        assertEquals("[]", items.toString())
        assertEquals(0, buffer.index.toLong())
        assertEquals(3, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertFalse(buffer.canUndo)
        assertFalse(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoMove() {
        testPushMove()

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoMoveDown() {
        testPushMoveDown()

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoMoveUp() {
        testPushMoveUp()

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoRemove() {
        testPushRemove()

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }

    @Test
    fun testUndoReplace() {
        testPushReplace()

        buffer.undo()
        assertEquals("[a, b, c]", items.toString())
        assertEquals(3, buffer.index.toLong())
        assertEquals(4, buffer.actions.size.toLong())
        assertTrue(buffer.canRedo)
        assertTrue(buffer.canRedoProperty().get())
        assertTrue(buffer.canUndo)
        assertTrue(buffer.canUndoProperty().get())
    }
}
